src/System/Random/Internal.hs

```diff
@@ -0,0 +1,1040 @@
+word32ToChar (W32# w#) = C# (chr# (word2Int# w#))
+{-# INLINE word32ToChar #-}
+
+charToWord32 :: Char -> Word32
+charToWord32 (C# c#) = W32# (int2Word# (ord# c#))
+{-# INLINE charToWord32 #-}
+
+instance Uniform Char where
+  uniformM g = word32ToChar <$> unbiasedWordMult32 (charToWord32 maxBound) g
+  {-# INLINE uniformM #-}
+instance UniformRange Char where
+  uniformRM (l, h) g =
+    word32ToChar <$> unbiasedWordMult32RM (charToWord32 l, charToWord32 h) g
+  {-# INLINE uniformRM #-}
+
+instance Uniform Bool where
+  uniformM = fmap wordToBool . uniformWord8
+    where wordToBool w = (w .&. 1) /= 0
+instance UniformRange Bool where
+  uniformRM (False, False) _g = return False
+  uniformRM (True, True)   _g = return True
+  uniformRM _               g = uniformM g
+
+-- | See [Floating point number caveats](System-Random-Stateful.html#fpcaveats).
+instance UniformRange Double where
+  uniformRM (l, h) g = do
+    w64 <- uniformWord64 g
+    let x = word64ToDoubleInUnitInterval w64
+    return $ (h - l) * x + l
+
+-- | Turns a given uniformly distributed 'Word64' value into a uniformly
+-- distributed 'Double' value in the range [0, 1].
+word64ToDoubleInUnitInterval :: Word64 -> Double
+word64ToDoubleInUnitInterval w64 = d / m
+  where
+    d = fromIntegral w64 :: Double
+    m = fromIntegral (maxBound :: Word64) :: Double
+{-# INLINE word64ToDoubleInUnitInterval #-}
+
+-- | See [Floating point number caveats](System-Random-Stateful.html#fpcaveats).
+instance UniformRange Float where
+  uniformRM (l, h) g = do
+    w32 <- uniformWord32 g
+    let x = word32ToFloatInUnitInterval w32
+    return $ (h - l) * x + l
+
+-- | Turns a given uniformly distributed 'Word32' value into a uniformly
+-- distributed 'Float' value in the range [0,1].
+word32ToFloatInUnitInterval :: Word32 -> Float
+word32ToFloatInUnitInterval w32 = f / m
+  where
+    f = fromIntegral w32 :: Float
+    m = fromIntegral (maxBound :: Word32) :: Float
+{-# INLINE word32ToFloatInUnitInterval #-}
+
+-- The two integer functions below take an [inclusive,inclusive] range.
+randomIvalIntegral :: (RandomGen g, Integral a) => (a, a) -> g -> (a, g)
+randomIvalIntegral (l,h) = randomIvalInteger (toInteger l, toInteger h)
+
+{-# SPECIALIZE randomIvalInteger :: (Num a) =>
+    (Integer, Integer) -> StdGen -> (a, StdGen) #-}
+
+randomIvalInteger :: (RandomGen g, Num a) => (Integer, Integer) -> g -> (a, g)
+randomIvalInteger (l,h) rng
+ | l > h     = randomIvalInteger (h,l) rng
+ | otherwise = case f 1 0 rng of (v, rng') -> (fromInteger (l + v `mod` k), rng')
+     where
+       (genlo, genhi) = genRange rng
+       b = fromIntegral genhi - fromIntegral genlo + 1 :: Integer
+
+       -- Probabilities of the most likely and least likely result
+       -- will differ at most by a factor of (1 +- 1/q). Assuming the RandomGen
+       -- is uniform, of course
+
+       -- On average, log q / log b more pseudo-random values will be generated
+       -- than the minimum
```

Comment:
On average by what? Minimum of what?

Reply:
Who in the hell knows :D 
Again, old deprecated function, let's leave it as is.